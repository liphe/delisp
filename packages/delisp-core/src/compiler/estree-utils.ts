import * as JS from "estree";
import { isValidJSIdentifierName } from "./jsvariable";

export function member(obj: JS.Expression, prop: string): JS.MemberExpression {
  const dotNotation = isValidJSIdentifierName(prop);
  return {
    type: "MemberExpression",
    computed: !dotNotation,
    object: obj,
    property: dotNotation
      ? { type: "Identifier", name: prop }
      : { type: "Literal", value: prop }
  };
}

export function methodCall(
  e: JS.Expression,
  method: string,
  args: JS.Expression[]
): JS.CallExpression {
  return {
    type: "CallExpression",
    callee: member(e, method),
    arguments: args
  };
}

export function arrowFunction(
  args: string[],
  body: JS.Expression
): JS.ArrowFunctionExpression {
  return {
    type: "ArrowFunctionExpression",
    params: args.map(name => ({ type: "Identifier", name })),
    body,
    generator: false,
    expression: true,
    async: false
  };
}

export function literal(value: number | string): JS.Literal {
  return {
    type: "Literal",
    value
  };
}

export function identifier(name: string): JS.Identifier {
  return {
    type: "Identifier",
    name
  };
}

export function primitiveCall(
  name: string,
  ...args: JS.Expression[]
): JS.Expression {
  return {
    type: "CallExpression",
    callee: identifier(name),
    arguments: args
  };
}
