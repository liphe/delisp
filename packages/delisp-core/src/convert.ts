import { printHighlightedExpr } from "./error-report";
import {
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";
import {
  Declaration,
  Expression,
  LambdaList,
  SIdentifier,
  SLetBinding,
  Syntax
} from "./syntax";
import { last } from "./utils";

import {
  checkUserDefinedTypeName,
  convert as convertType
} from "./convert-type";
import { parseRecord } from "./convert-utils";
import { listTypeVariables } from "./type-utils";
import { TypeWithWildcards } from "./type-wildcards";

const conversions: Map<string, (expr: ASExprList) => Expression> = new Map();
const toplevelConversions: Map<
  string,
  (expr: ASExprList) => Declaration
> = new Map();

function defineConversion(name: string, fn: (expr: ASExprList) => Expression) {
  conversions.set(name, fn);
}

function defineToplevel(name: string, fn: (expr: ASExprList) => Declaration) {
  toplevelConversions.set(name, fn);
}

function parseBody(anchor: ASExpr, exprs: ASExpr[]): Expression[] {
  if (exprs.length === 0) {
    throw new Error(
      printHighlightedExpr(`body can't be empty`, anchor.location, true)
    );
  }
  return exprs.map(convertExpr);
}

//
// The format of lambda lists are (a b c ... &rest z)
//

function parseLambdaList(ll: ASExpr): LambdaList {
  if (ll.tag !== "list") {
    throw new Error(
      printHighlightedExpr("Expected a list of arguments", ll.location)
    );
  }

  ll.elements.forEach(arg => {
    if (arg.tag !== "symbol") {
      throw new Error(
        printHighlightedExpr(
          "A list of arguments should be made of symbols",
          arg.location
        )
      );
    }
  });

  const symbols = ll.elements as ASExprSymbol[];

  // Check for duplicated arguments
  symbols.forEach((arg, i) => {
    const duplicated = symbols.slice(i + 1).find(a => a.name === arg.name);
    if (duplicated) {
      throw new Error(
        printHighlightedExpr(
          "There is another argument with the same name",
          duplicated.location
        )
      );
    }
  });

  return {
    positionalArgs: symbols.map(convertSymbol),
    location: ll.location
  };
}

defineConversion("lambda", expr => {
  const [lambda, ...args] = expr.elements;
  const lastExpr = last([lambda, ...args]) as ASExpr; // we kj

  if (args.length === 0) {
    throw new Error(
      printHighlightedExpr(
        `'lambda' is missing the argument list`,
        lastExpr.location,
        true
      )
    );
  }

  const body = parseBody(lastExpr, args.slice(1));

  return {
    tag: "function",
    lambdaList: parseLambdaList(args[0]),
    body,
    location: expr.location,
    info: {}
  };
});

defineConversion("if", expr => {
  if (expr.elements.length !== 4) {
    const lastExpr = last(expr.elements) as ASExpr; // we know it is not empty!
    throw new Error(
      printHighlightedExpr(
        `'if' needs exactly 3 arguments, got ${expr.elements.length}`,
        lastExpr.location,
        true
      )
    );
  }
  const [, conditionForm, consequentForm, alternativeForm] = expr.elements;
  return {
    tag: "conditional",
    condition: convertExpr(conditionForm),
    consequent: convertExpr(consequentForm),
    alternative: convertExpr(alternativeForm),
    location: expr.location,
    info: {}
  };
});

function parseLetBindings(bindings: ASExpr): SLetBinding[] {
  if (bindings.tag !== "map") {
    throw new Error(
      printHighlightedExpr(`'let' bindings should be a map`, bindings.location)
    );
  }

  return bindings.fields.map(field => ({
    variable: convertSymbol(field.label),
    value: convertExpr(field.value)
  }));
}

defineConversion("let", expr => {
  const [_let, ...args] = expr.elements;
  const lastExpr = last([_let, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new Error(
      printHighlightedExpr(
        `'let' is missing the bindings`,
        lastExpr.location,
        true
      )
    );
  }

  const [rawBindings, ...rawBody] = args;

  return {
    tag: "let-bindings",
    bindings: parseLetBindings(rawBindings),
    body: parseBody(lastExpr, rawBody),
    location: expr.location,
    info: {}
  };
});

defineConversion("the", expr => {
  const [_the, ...args] = expr.elements;
  const lastExpr = last([_the, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new Error(
      printHighlightedExpr(
        `'the' is missing the type and value`,
        lastExpr.location,
        true
      )
    );
  }
  if (args.length === 1) {
    throw new Error(
      printHighlightedExpr(
        `'the' is missing the expression`,
        lastExpr.location,
        true
      )
    );
  }

  if (args.length > 2) {
    throw new Error(
      printHighlightedExpr(
        `Too many arguments. 'the' should take two arguments.`,
        args[2].location,
        true
      )
    );
  }

  const [t, value] = args;

  return {
    tag: "type-annotation",
    typeWithWildcards: new TypeWithWildcards(convertType(t)),
    value: convertExpr(value),
    location: expr.location,
    info: {}
  };
});

defineToplevel("define", expr => {
  const [define, ...args] = expr.elements;

  if (args.length !== 2) {
    const lastExpr = last([define, ...args]) as ASExpr;
    throw new Error(
      printHighlightedExpr(
        `'define' needs exactly 2 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [variable, value] = args;

  if (variable.tag !== "symbol") {
    throw new Error(
      printHighlightedExpr("'define' expected a symbol", variable.location)
    );
  }

  return {
    tag: "definition",
    variable: convertSymbol(variable),
    value: convertExpr(value),
    location: expr.location
  };
});

defineToplevel("export", expr => {
  const [exp, ...args] = expr.elements;

  if (args.length !== 1) {
    const lastExpr = last([exp, ...args]) as ASExpr;
    throw new Error(
      printHighlightedExpr(
        `'export' needs exactly 1 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [variable] = args;

  if (variable.tag !== "symbol") {
    throw new Error(
      printHighlightedExpr("'export' expected a symbol", variable.location)
    );
  }

  return {
    tag: "export",
    value: convertSymbol(variable),
    location: expr.location
  };
});

defineToplevel("type", expr => {
  const [typeOp, ...args] = expr.elements;

  if (args.length !== 2) {
    const lastExpr = last([typeOp, ...args]) as ASExpr;
    throw new Error(
      printHighlightedExpr(
        `'type' needs exactly 2 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [name, definition] = args;

  if (name.tag !== "symbol") {
    throw new Error(
      printHighlightedExpr("'type' expected a symbol as a name", name.location)
    );
  }

  checkUserDefinedTypeName(name);

  const definitionType = convertType(definition);
  if (listTypeVariables(definitionType).length > 0) {
    throw new Error(
      printHighlightedExpr("Type variable out of scope", definition.location)
    );
  }

  return {
    tag: "type-alias",
    alias: convertSymbol(name),
    definition: definitionType,
    location: expr.location
  };
});

function convertList(list: ASExprList): Expression {
  if (list.elements.length === 0) {
    throw new Error(
      printHighlightedExpr("Empty list is not a function call", list.location)
    );
  }

  const [first] = list.elements;

  const convertSpecialForm =
    first.tag === "symbol" ? conversions.get(first.name) : undefined;

  if (convertSpecialForm) {
    return convertSpecialForm(list);
  } else {
    const [fn, ...args] = list.elements;
    return {
      tag: "function-call",
      fn: convertExpr(fn),
      args: args.map(convertExpr),
      location: list.location,
      info: {}
    };
  }
}

function convertVector(list: ASExprVector): Expression {
  return {
    tag: "vector",
    values: list.elements.map(a => convertExpr(a)),
    location: list.location,
    info: {}
  };
}

function convertMap(map: ASExprMap): Expression {
  const { fields, tail } = parseRecord(map);

  return {
    tag: "record",
    fields: fields.map(f => ({
      label: convertSymbol(f.label),
      value: convertExpr(f.value)
    })),
    extends: tail && convertExpr(tail),
    location: map.location,
    info: {}
  };
}

function convertSymbol(expr: ASExprSymbol): SIdentifier {
  return {
    tag: "identifier",
    name: expr.name,
    location: expr.location,
    info: {}
  };
}

export function convertExpr(expr: ASExpr): Expression {
  switch (expr.tag) {
    case "number":
      return { ...expr, info: {} };
    case "string":
      return { ...expr, info: {} };
    case "symbol":
      return convertSymbol(expr);
    case "list":
      return convertList(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
  }
}

export function convert(expr: ASExpr): Syntax {
  if (expr.tag === "list") {
    if (expr.elements.length === 0) {
      throw new Error(
        printHighlightedExpr("Empty list is not a function call", expr.location)
      );
    }

    const [first] = expr.elements;
    const convertDeclaration =
      first.tag === "symbol" ? toplevelConversions.get(first.name) : undefined;

    if (convertDeclaration) {
      return convertDeclaration(expr);
    }
  }

  return convertExpr(expr);
}
