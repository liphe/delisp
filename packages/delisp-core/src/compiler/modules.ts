import * as JS from "estree";
import { member } from "./estree-utils";
import { identifierToJS } from "./jsvariable";

import { printHighlightedExpr } from "../error-report";
import { SExport } from "../syntax";

export interface ModuleBackend {
  export(
    name: string,
    value: JS.Expression,
    exp: SExport
  ): JS.Statement | JS.ModuleDeclaration;
}

export const cjs: ModuleBackend = {
  export(name, value) {
    return {
      type: "ExpressionStatement",
      expression: {
        type: "AssignmentExpression",
        operator: "=",
        left: member(
          {
            type: "MemberExpression",
            computed: false,
            object: { type: "Identifier", name: "module" },
            property: { type: "Identifier", name: "exports" }
          },
          identifierToJS(name)
        ),
        right: value
      }
    };
  }
};

export const esm: ModuleBackend = {
  export(name, value, exp) {
    if (value.type !== "Identifier") {
      throw new Error(
        printHighlightedExpr(
          "Only user defined symbols can be exported",
          exp.value.location
        )
      );
    }
    return {
      type: "ExportNamedDeclaration",
      exportKind: "value",
      specifiers: [
        {
          type: "ExportSpecifier",
          local: value,
          exported: { type: "Identifier", name: identifierToJS(name) }
        }
      ],
      declaration: null
    };
  }
};
