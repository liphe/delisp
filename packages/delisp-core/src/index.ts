import { convert as convertSyntax } from "./convert";

import { readFromString } from "./reader";

export { collectConvertErrors } from "./convert";

export {
  compileToString,
  compileModuleToString,
  moduleEnvironment
} from "./compiler";
export { createContext, evaluate, evaluateModule } from "./eval";
export {
  inferType,
  inferModule,
  inferExpressionInModule,
  defaultEnvironment
} from "./infer";
export { Type } from "./types";
export { printType } from "./type-printer";
export {
  isDeclaration,
  isDefinition,
  isExpression,
  isExport,
  isTypeAlias
} from "./syntax";
export { default as primitives } from "./primitives";

export { printHighlightedExpr } from "./error-report";

export function readSyntax(source: string) {
  return convertSyntax(readFromString(source));
}

export { findSyntaxByOffset, findSyntaxByRange } from "./syntax-utils";

export {
  createModule,
  readModule,
  addToModule,
  removeModuleDefinition,
  removeModuleTypeDefinition,
  moduleDefinitions,
  moduleDefinitionByName
} from "./module";

export { generateTSModuleDeclaration } from "./typescript-generation";

export { Encoder } from "./prettier";
export { pprintAs, pprint, pprintModuleAs, pprintModule } from "./printer";
