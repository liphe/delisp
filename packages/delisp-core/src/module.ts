import { WithErrors, convert } from "./convert";
import { readAllFromString } from "./reader";
import { Module, Syntax, SDefinition, isDefinition } from "./syntax";

export function createModule(): Module {
  return {
    tag: "module",
    body: []
  };
}

export function readModule(str: string): Module<WithErrors, WithErrors> {
  return {
    tag: "module",
    body: readAllFromString(str).map(convert)
  };
}

export function addToModule(m: Module, s: Syntax): Module {
  return {
    tag: "module",
    body: [...m.body, s]
  };
}

export function removeModuleDefinition(m: Module, name: string): Module {
  return {
    tag: "module",
    body: m.body.filter(d => {
      return d.node.tag === "definition" ? d.node.variable.name !== name : true;
    })
  };
}

export function removeModuleTypeDefinition(m: Module, name: string): Module {
  return {
    tag: "module",
    body: m.body.filter(d => {
      return d.node.tag === "type-alias" ? d.node.alias.name !== name : true;
    })
  };
}

export function moduleDefinitions<A>(m: Module<A>): Array<SDefinition<A>> {
  return m.body.filter(isDefinition);
}

export function moduleDefinitionByName<A>(
  name: string,
  m: Module<A>
): SDefinition<A> | null {
  return (
    moduleDefinitions(m).find(def => def.node.variable.name === name) || null
  );
}
