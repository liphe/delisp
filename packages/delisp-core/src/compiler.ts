import runtime from "@delisp/runtime";
import {
  Expression,
  isDefinition,
  isExport,
  isTypeAlias,
  Module,
  SConditional,
  SDefinition,
  SExport,
  SFunction,
  SFunctionCall,
  SIdentifier,
  SLet,
  SRecord,
  SVectorConstructor,
  Syntax
} from "./syntax";

import { methodCall } from "./compiler/estree-utils";
import { last, mapObject, maybeMap } from "./utils";

import { InvariantViolation } from "./invariant";

import { printHighlightedExpr } from "./error-report";

import {
  DefinitionBackend,
  dynamicDefinition,
  staticDefinition
} from "./compiler/definitions";
import { cjs, esm, ModuleBackend } from "./compiler/modules";

import { member } from "./compiler/estree-utils";
import {
  compileInlinePrimitive,
  isInlinePrimitive
} from "./compiler/inline-primitives";
import { identifierToJS, isValidJSIdentifierName } from "./compiler/jsvariable";
import { pprint } from "./printer";

import * as escodegen from "escodegen";
import * as JS from "estree";

import createDebug from "debug";
const debug = createDebug("delisp:compiler");

interface EnvironmentBinding {
  jsname: string;
  source: "lexical" | "module" | "primitive";
}

export interface Environment {
  defs: DefinitionBackend;
  moduleFormat: ModuleBackend;
  bindings: {
    [symbol: string]: EnvironmentBinding;
  };
}

export interface CompilerOptions {
  definitionContainer?: string;
  esModule?: boolean;
}

function addBinding(varName: string, env: Environment): Environment {
  return {
    ...env,
    bindings: {
      ...env.bindings,
      [varName]: { jsname: identifierToJS(varName), source: "lexical" }
    }
  };
}

function lookupBinding(varName: string, env: Environment) {
  return env.bindings[varName];
}

function compileBody(body: Expression[], env: Environment): JS.BlockStatement {
  const middleForms = body.slice(0, -1);
  const lastForm = last(body)!;

  return {
    type: "BlockStatement",
    body: [
      ...middleForms.map(
        (e): JS.ExpressionStatement => ({
          type: "ExpressionStatement",
          expression: compile(e, env)
        })
      ),
      {
        type: "ReturnStatement",
        argument: compile(lastForm, env)
      }
    ]
  };
}

// Convert a Delisp variable name to Javascript. This function should
// be injective so there is no collisions and the output should be a
// valid variable name.

function compileLambda(
  fn: SFunction,
  env: Environment
): JS.ArrowFunctionExpression {
  const newEnv = fn.lambdaList.positionalArgs.reduce(
    (e, param) => addBinding(param.name, e),
    env
  );

  const jsargs = fn.lambdaList.positionalArgs.map(
    (param): JS.Pattern => ({
      type: "Identifier",
      name: lookupBinding(param.name, newEnv).jsname
    })
  );

  const implicitReturn = fn.body.length === 1;

  const body: JS.Expression | JS.Statement = implicitReturn
    ? compile(fn.body[0], newEnv)
    : compileBody(fn.body, newEnv);

  return {
    type: "ArrowFunctionExpression",
    params: [...jsargs],
    body,
    expression: implicitReturn
  };
}

function compileDefinition(def: SDefinition, env: Environment): JS.Statement {
  const value = compile(def.value, env);
  const name = lookupBinding(def.variable, env).jsname;
  return env.defs.define(name, value);
}

function compileFunctionCall(
  funcall: SFunctionCall,
  env: Environment
): JS.Expression {
  const compiledArgs = funcall.args.map(arg => compile(arg, env));
  if (funcall.fn.tag === "identifier" && isInlinePrimitive(funcall.fn.name)) {
    return compileInlinePrimitive(funcall.fn.name, compiledArgs, "funcall");
  } else {
    return {
      type: "CallExpression",
      callee: compile(funcall.fn, env),
      arguments: compiledArgs
    };
  }
}

function compileIdentifier(ref: SIdentifier, env: Environment): JS.Expression {
  const binding = lookupBinding(ref.name, env);

  if (!binding) {
    if (isInlinePrimitive(ref.name)) {
      return compileInlinePrimitive(ref.name, [], "value");
    } else {
      return env.defs.access(identifierToJS(ref.name));
    }
  }

  switch (binding.source) {
    case "primitive":
      return member(
        {
          type: "Identifier",
          name: "env"
        },
        binding.jsname
      );
    case "module":
      return env.defs.access(binding.jsname);
    case "lexical":
      return {
        type: "Identifier",
        name: binding.jsname
      };
    default:
      throw new InvariantViolation(
        "This switch-statement should be exhaustive but TS doesn't detect it somehow."
      );
  }
}

function compileConditional(
  expr: SConditional,
  env: Environment
): JS.Expression {
  return {
    type: "ConditionalExpression",
    test: compile(expr.condition, env),
    consequent: compile(expr.consequent, env),
    alternate: compile(expr.alternative, env)
  };
}

function compileLetBindings(expr: SLet, env: Environment): JS.Expression {
  const newenv = expr.bindings.reduce(
    (e, binding) => addBinding(binding.var, e),
    env
  );

  return {
    type: "CallExpression",
    callee: {
      type: "FunctionExpression",
      params: expr.bindings.map(
        (b): JS.Pattern => ({
          type: "Identifier",
          name: lookupBinding(b.var, newenv).jsname
        })
      ),
      body: compileBody(expr.body, newenv)
    },
    arguments: expr.bindings.map(b => compile(b.value, env))
  };
}

function literal(value: number | string): JS.Literal {
  return {
    type: "Literal",
    value
  };
}

function identifier(name: string): JS.Identifier {
  return {
    type: "Identifier",
    name
  };
}

function compileVector(
  expr: SVectorConstructor,
  env: Environment
): JS.Expression {
  return {
    type: "ArrayExpression",
    elements: expr.values.map(e => compile(e, env))
  };
}

function compileRecord(expr: SRecord, env: Environment): JS.Expression {
  const newObj: JS.ObjectExpression = {
    type: "ObjectExpression",
    properties: expr.fields.map(
      ({ label, value }): JS.Property => {
        if (!label.startsWith(":")) {
          throw new InvariantViolation(`Invalid record ${label}`);
        }

        const name = label.replace(/^:/, "");

        return {
          type: "Property",
          key: isValidJSIdentifierName(name) ? identifier(name) : literal(name),
          value: compile(value, env),
          kind: "init",
          method: false,
          shorthand: false,
          computed: false
        };
      }
    )
  };
  if (expr.extends) {
    return methodCall({ type: "Identifier", name: "Object" }, "assign", [
      { type: "ObjectExpression", properties: [] },
      compile(expr.extends, env),
      newObj
    ]);
  } else {
    return newObj;
  }
}

function compileNumber(value: number): JS.Expression {
  if (value >= 0) {
    return literal(value);
  } else {
    return {
      type: "UnaryExpression",
      operator: "-",
      prefix: true,
      argument: literal(-value)
    };
  }
}

export function compile(expr: Expression, env: Environment): JS.Expression {
  switch (expr.tag) {
    case "number":
      return compileNumber(expr.value);
    case "string":
      return literal(expr.value);
    case "vector":
      return compileVector(expr, env);
    case "record":
      return compileRecord(expr, env);
    case "identifier":
      return compileIdentifier(expr, env);
    case "conditional":
      return compileConditional(expr, env);
    case "function":
      return compileLambda(expr, env);
    case "function-call":
      return compileFunctionCall(expr, env);
    case "let-bindings":
      return compileLetBindings(expr, env);
    case "type-annotation":
      return compile(expr.value, env);
  }
}

function compileTopLevel(
  syntax: Syntax,
  env: Environment
): JS.Statement | null {
  if (isExport(syntax) || isTypeAlias(syntax)) {
    // exports are compiled at the end of the module
    return null;
  }

  let js: JS.Statement;

  if (isDefinition(syntax)) {
    js = compileDefinition(syntax, env);
  } else {
    js = {
      type: "ExpressionStatement",
      expression: compile(syntax, env)
    };
  }

  return {
    ...js,
    // Include a comment with the original source code immediately
    // before each toplevel compilation.
    leadingComments: [
      {
        type: "Block",
        value: `
${pprint(syntax, 60)}
`
      }
    ]
  };
}

function compileRuntime(env: Environment): JS.Statement | JS.ModuleDeclaration {
  return env.moduleFormat.importRuntime("env");
}

function compileExports(
  exps: SExport[],
  env: Environment
): Array<JS.Statement | JS.ModuleDeclaration> {
  const exportNames = exps.map(exp => {
    const binding = lookupBinding(exp.value.name, env);
    if (!binding || binding.source !== "module") {
      throw new Error(
        printHighlightedExpr(
          "You can only export user definitions",
          exp.value.location
        )
      );
    } else {
      return binding.jsname;
    }
  });

  if (exportNames.length > 0) {
    return [env.moduleFormat.export(exportNames)];
  } else {
    return [];
  }
}

export function moduleEnvironment(
  m: Module,
  opts: CompilerOptions = {}
): Environment {
  const moduleDefinitions = m.body
    .filter(isDefinition)
    .map(decl => decl.variable);
  const moduleBindings = moduleDefinitions.reduce(
    (d, decl) => ({
      ...d,
      [decl]: { jsname: identifierToJS(decl), source: "module" }
    }),
    {}
  );

  const primitiveBindings = mapObject(
    runtime,
    (_, key: string): EnvironmentBinding => ({
      jsname: key,
      source: "primitive"
    })
  );

  const initialEnv = {
    defs: opts.definitionContainer
      ? dynamicDefinition(opts.definitionContainer)
      : staticDefinition,
    moduleFormat: opts.esModule ? esm : cjs,
    bindings: { ...primitiveBindings, ...moduleBindings }
  };

  return initialEnv;
}

function compileModule(
  m: Module,
  includeRuntime: boolean,
  env: Environment
): JS.Program {
  return {
    type: "Program",
    sourceType: "module",
    body: [
      ...(includeRuntime ? [compileRuntime(env)] : []),
      ...maybeMap((syntax: Syntax) => compileTopLevel(syntax, env), m.body),
      ...compileExports(m.body.filter(isExport), env)
    ]
  };
}

export function compileToString(syntax: Syntax, env: Environment): string {
  const ast = compileModule({ tag: "module", body: [syntax] }, false, env);
  const code = escodegen.generate(ast, { comment: true });
  debug("jscode:", code);
  return code;
}

export function compileModuleToString(
  m: Module,
  opts: CompilerOptions = {}
): string {
  const env = moduleEnvironment(m, opts);
  const ast = compileModule(m, true, env);
  const code = escodegen.generate(ast, { comment: true });
  debug("jscode:", code);
  return code;
}
