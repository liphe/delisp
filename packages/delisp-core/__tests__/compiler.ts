import { readFromString } from "../src/reader";
import { convert } from "../src/convert";
import { createModule } from "../src/module";
import { compileToString, moduleEnvironment } from "../src/compiler";

describe("Compiler", () => {
  describe("Error messages", () => {
    function compileError(str: string): string {
      let result: string | undefined;
      try {
        const sexpr = readFromString(str);
        const syntax = convert(sexpr);
        const env = moduleEnvironment(createModule());
        compileToString(syntax, env);
      } catch (err) {
        result = err.message;
      }
      if (result) {
        return result;
      } else {
        throw new Error(`FATAL: EXPRESSION DID NOT FAIL TO COMPILE.`);
      }
    }

    it("generate nice error message for invalid lambda expressions", () => {
      expect(compileError("(lambda)")).toMatchSnapshot();
      expect(compileError("(lambda 7)")).toMatchSnapshot();
      expect(compileError("(lambda 7 5)")).toMatchSnapshot();
      expect(compileError("(lambda (7) x)")).toMatchSnapshot();
      expect(compileError("(lambda (x 7) x)")).toMatchSnapshot();
    });

    it("generate nice error message for invalid let expressions", () => {
      expect(compileError("(let)")).toMatchSnapshot();
      expect(compileError("(let x 5)")).toMatchSnapshot();
      expect(compileError("(let (x 5) x)")).toMatchSnapshot();
      expect(compileError("(let {x} x)")).toMatchSnapshot();
      expect(compileError("(let {a b c} x)")).toMatchSnapshot();
    });

    it("generate nice error message for invalid definitions", () => {
      expect(compileError("(define)")).toMatchSnapshot();
      expect(compileError("(define 5)")).toMatchSnapshot();
      expect(compileError("(define x)")).toMatchSnapshot();
      expect(compileError("(define 4 2)")).toMatchSnapshot();
    });

    it("generate nice error message for conditionals", () => {
      expect(compileError("(if)")).toMatchSnapshot();
      expect(compileError("(if 5)")).toMatchSnapshot();
      expect(compileError("(if x 2)")).toMatchSnapshot();
      expect(compileError("(if x 2 3 0)")).toMatchSnapshot();
    });

    it("generate nice error message for invalid records", () => {
      expect(compileError("{:x 10 | y :z 20}")).toMatchSnapshot();
    });

    it("generate nice error message for records with duplicated labels", () => {
      expect(compileError("{:x 10 :x 20}")).toMatchSnapshot();
    });
  });
});
