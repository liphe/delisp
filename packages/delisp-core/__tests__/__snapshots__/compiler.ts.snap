// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler Error messages generate a nice error for some basic invalid syntax 1`] = `
"file:1:1: Empty list is not a function call
()
^"
`;

exports[`Compiler Error messages generate nice error message for bad type declarations 1`] = `
"file:1:5: 'type' needs exactly 2 arguments, got 0
(type)
----^"
`;

exports[`Compiler Error messages generate nice error message for bad type declarations 2`] = `
"file:1:7: 'type' needs exactly 2 arguments, got 1
(type a)
------^"
`;

exports[`Compiler Error messages generate nice error message for bad type declarations 3`] = `
"file:1:7: 'type' needs exactly 2 arguments, got 1
(type A)
------^"
`;

exports[`Compiler Error messages generate nice error message for bad type declarations 4`] = `
"file:1:9: Not a valid type
(type A 3)
--------^"
`;

exports[`Compiler Error messages generate nice error message for bad type declarations 5`] = `
"file:1:7: 'a' is not a valid name for a type. Type names should start with a capital letter. Try 'A'?
(type a {})
------^"
`;

exports[`Compiler Error messages generate nice error message for bad type declarations 6`] = `
"file:1:14: 'type' needs exactly 2 arguments, got 3
(type ID {} 10)
-------------^"
`;

exports[`Compiler Error messages generate nice error message for bad type declarations 7`] = `
"file:1:7: 'type' expected a symbol as a name
(type 3 5)
------^"
`;

exports[`Compiler Error messages generate nice error message for bad type declarations 8`] = `
"file:1:10: Type variable out of scope
(type ID (-> a a))
---------^"
`;

exports[`Compiler Error messages generate nice error message for conditionals 1`] = `
"file:1:3: 'if' needs exactly 3 arguments, got 1
(if)
--^"
`;

exports[`Compiler Error messages generate nice error message for conditionals 2`] = `
"file:1:5: 'if' needs exactly 3 arguments, got 2
(if 5)
----^"
`;

exports[`Compiler Error messages generate nice error message for conditionals 3`] = `
"file:1:7: 'if' needs exactly 3 arguments, got 3
(if x 2)
------^"
`;

exports[`Compiler Error messages generate nice error message for conditionals 4`] = `
"file:1:11: 'if' needs exactly 3 arguments, got 5
(if x 2 3 0)
----------^"
`;

exports[`Compiler Error messages generate nice error message for invalid definitions 1`] = `
"file:1:7: 'define' needs exactly 2 arguments, got 0
(define)
------^"
`;

exports[`Compiler Error messages generate nice error message for invalid definitions 2`] = `
"file:1:9: 'define' needs exactly 2 arguments, got 1
(define 5)
--------^"
`;

exports[`Compiler Error messages generate nice error message for invalid definitions 3`] = `
"file:1:9: 'define' needs exactly 2 arguments, got 1
(define x)
--------^"
`;

exports[`Compiler Error messages generate nice error message for invalid definitions 4`] = `
"file:1:9: 'define' expected a symbol
(define 4 2)
--------^"
`;

exports[`Compiler Error messages generate nice error message for invalid definitions 5`] = `
"file:1:15: 'define' needs exactly 2 arguments, got 3
(define x 10 23)
--------------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 1`] = `
"file:1:7: 'lambda' is missing the argument list
(lambda)
------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 2`] = `
"file:1:9: body can't be empty
(lambda 7)
--------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 3`] = `
"file:1:9: Expected a list of arguments
(lambda 7 5)
--------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 4`] = `
"file:1:10: A list of arguments should be made of symbols
(lambda (7) x)
---------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 5`] = `
"file:1:12: A list of arguments should be made of symbols
(lambda (x 7) x)
-----------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 6`] = `
"file:1:12: There is another argument with the same name
(lambda (x x) x)
-----------^"
`;

exports[`Compiler Error messages generate nice error message for invalid let expressions 1`] = `
"file:1:4: 'let' is missing the bindings
(let)
---^"
`;

exports[`Compiler Error messages generate nice error message for invalid let expressions 2`] = `
"file:1:6: 'let' bindings should be a map
(let x 5)
-----^"
`;

exports[`Compiler Error messages generate nice error message for invalid let expressions 3`] = `
"file:1:6: 'let' bindings should be a map
(let (x 5) x)
-----^"
`;

exports[`Compiler Error messages generate nice error message for invalid let expressions 4`] = `
"file:1:9: Expected an alphanumeric character
(let {x} x)
--------^"
`;

exports[`Compiler Error messages generate nice error message for invalid let expressions 5`] = `
"file:1:13: Expected an alphanumeric character
(let {a b c} x)
------------^"
`;

exports[`Compiler Error messages generate nice error message for invalid records 1`] = `
"file:1:8: '|' is not a valid field name. Fields should start with ':' (colon).
{:x 10 | y :z 20}
-------^"
`;

exports[`Compiler Error messages generate nice error message for invalid type annotations 1`] = `
"file:1:4: 'the' is missing the type and value
(the)
---^"
`;

exports[`Compiler Error messages generate nice error message for invalid type annotations 2`] = `
"file:1:6: 'the' is missing the expression
(the 3)
-----^"
`;

exports[`Compiler Error messages generate nice error message for invalid type annotations 3`] = `
"file:1:7: 'the' is missing the expression
(the ID)
------^"
`;

exports[`Compiler Error messages generate nice error message for invalid type annotations 4`] = `
"file:1:6: Not a valid type
(the 3 3)
-----^"
`;

exports[`Compiler Error messages generate nice error message for invalid type annotations 5`] = `
"file:1:11: Too many arguments. 'the' should take two arguments.
(the ID 3 3)
----------^"
`;

exports[`Compiler Error messages generate nice error message for records with duplicated labels 1`] = `
"file:1:8: Duplicated label
{:x 10 :x 20}
-------^"
`;

exports[`Compiler Error messages generate nice error messages for invalid exports 1`] = `
"file:1:7: 'export' needs exactly 1 arguments, got 0
(export)
------^"
`;

exports[`Compiler Error messages generate nice error messages for invalid exports 2`] = `
"file:1:9: 'export' expected a symbol
(export 1)
--------^"
`;

exports[`Compiler Error messages generate nice error messages for invalid exports 3`] = `
"file:1:9: 'export' expected a symbol
(export (+ 1 2))
--------^"
`;

exports[`Compiler Error messages generate nice error messages for invalid exports 4`] = `
"file:1:13: 'export' needs exactly 1 arguments, got 3
(export 1 2 3)
------------^"
`;
